File: compiler/codegen.clj
(ns compiler.codegen
  (:require [clojure.set])
  (:use [compiler.ast])
  (:use [compiler.symbol])  
  (:use [compiler.bridge])
  (:use [compiler.error])  
  (:use [compiler.string2ast :only [xe]]))

; code generation
(defn interval [n m] ; returns the list (n n+1 n+2 ... m)
  (if (<= n m) (cons-clj n (interval (+ n 1) m)) (list)))

(def code-gen)
(def lambda-todo (atom (list)))
(def lambda-count (atom 0))

(defn add-lambda! [lam]
  (let [i @lambda-count]
    (swap! lambda-count #(+ % 1))
    (swap! lambda-todo #(cons-clj (cons-clj i lam) %))
    i))

(defn cg-list [asts vars stack-env sep cont]
  (if (empty? asts)
    (cont "" stack-env)
    (let [x (code-gen (car asts) stack-env)]
      (cg-list (cdr asts)
               (cdr vars)
               (cons-clj (car vars) stack-env)
               sep
               (fn [code stack-env]
                       (cont (list x sep code)
                             stack-env))))))

(defn cg-args [args stack-env]
  (cg-list args
           (interval 1 (count args))
           stack-env
           ""
           (fn [code stack-env]
                   code)))

(def global-vars (atom (list)))

(defn access-var [var stack-env]
  (if (global-var? var)
    (let [i (pos-in-list var @global-vars)]
      (list "GLOBAL(" i "/*" (var-uid var) "*/)"))
    (let [i (- (count stack-env)
                (pos-in-list var stack-env)
                1)]
      (list "LOCAL(" i "/*" (var-uid var) "*/)"))))

(defn cg [stack-env ast]

  (cond
    (lit? ast)
    (let [val (lit-val ast)]
               (case val
                 false (list " PUSH(FALSEOBJ));")
                 true (list " PUSH(TRUEOBJ));")
                 (list " PUSH(INT2OBJ(" val "));")))

    (ref? ast)
    (let [var (ref-var ast)]
      (list " PUSH(" (access-var var stack-env) ");"))

    (set-clj? ast)
    (let [var (set-var ast)]
      (list
       (cg stack-env (car (ast-subx ast)))
       " " (access-var var stack-env) " = TOS();"))

    (cnd? ast)
    (let [x (map (fn [_x] (cg stack-env _x)) (ast-subx ast))]
      (list (car x)
            "\n if (POP()) {\n"
            (cadr x)
            "\n } else {\n"
            (caddr x)
            "\n }"))

    (prim? ast)
    (let [args (ast-subx ast)
          op (prim-op ast)]
      
      (cond

        (= op '%=) (list (cg-args args stack-env) " EQ();")
        (= op '%<) (list (cg-args args stack-env) " LT();")
        (= op '%+) (list (cg-args args stack-env) " ADD();")
        (= op '%-) (list (cg-args args stack-env) " SUB();")
        (= op '%*) (list (cg-args args stack-env) " MUL();")
        (= op '%display) (list (cg-args args stack-env) " DISPLAY();")
        (= op '%halt) (list (cg-args args stack-env) " HALT();")	 

        (= op '%closure)
        (let [i (add-lambda! (car args))
              n (count (cdr args))
              s (list "CLOSURE(" i "," n ");")]
          (list
           (cg-args (cdr args) stack-env)
           " BEGIN_" s
           (map (fn [j]
                        (list " INICLO(" j ");"))
                (reverse (interval 1 n)))
           " END_" s))
      
        (= op '%closure-ref)
        (let [i (lit-val (cadr args))]
          (list
           (cg stack-env (car args))
           " TOS() = CLOSURE_REF(TOS()," i ");"))
      
        :else (error (str "unknown primitive " (= '%closure op)))))

    (app? ast)
    (let [fnc (car (ast-subx ast))
          args (cdr (ast-subx ast))
          n (count args)]
      (if (lam? fnc)
        (cg-list args
                 (lam-params fnc)
                 stack-env
                 "\n"
                 (fn [code new-stack-env]
                         (list
                          code
                          (code-gen (car (ast-subx fnc))
                                    new-stack-env))))
        (cg-list args
                 (interval 1 n)
                 stack-env
                 "\n"
                 (fn [code new-stack-env]
                   (let [start (count stack-env)
                         s (list "JUMP(" n ");")]
                     (list
                      code
                      " BEGIN_" s
                      (map (fn [j]
                             (list " PUSH(LOCAL(" (+ j start) "));"))
                           (interval 0 (- n 1)))
                      " END_" s))))))
  
    (lam? ast)
                                        ; this case is impossible after CPS-conversion
    (list " PUSH(INT2OBJ(" (add-lambda! ast) "));")
  
    (seq? ast)
                                        ; this case is impossible after CPS-conversion
    (map (fn [ast]
           (list (cg stack-env ast) "DROP();"))
         (ast-subx ast))

    :else
    (error (str "unknown ast" ast))))

    
(defn code-gen [ast stack-env]
    (cg stack-env ast))

(defn limit [s n]
  (let [c (count s)
        [i e] (if (> c n) [n "..."] [c ""])
        ]
    (str (subs s 0 i) e)))

(def source)
(defn compile-all-lambdas []
  (if (empty? @lambda-todo)
    ""
    (let [x (car @lambda-todo)
          ast (cdr x)]
      (swap! lambda-todo #(cdr %))
      (list
       "case " (car x) ": /* " (limit (pr-str (source ast)) 57) " */\n\n"
       (code-gen (car (ast-subx ast))
                 (reverse (lam-params ast)))
       "\n\n"
       (compile-all-lambdas)))))

(defn code-generate [ast]
  (let [total (slurp "compiler/runtime.c")
        m (re-matches #"(?s)(.*)//__SCHEME_CODE__.(.*)" total)
        prefix (m 1)
        suffix (m 2)]
   (swap! global-vars (fn [_x] (fv ast)))
   (add-lambda! ast)
   (let [code (compile-all-lambdas)]
       (list
        (list
 	"#define NB_GLOBALS " (count @global-vars) "\n"
 	"#define MAX_STACK " 100 "\n\n" ; could be computed...
 	prefix)
        code
        suffix))))


(defn source [ast]
  (cond (lit? ast)
        (lit-val ast)

        (ref? ast)
        (var-uid (ref-var ast))

        (set-clj? ast)
        (list 'set!
              (var-uid (set-var ast))
              (source (car (ast-subx ast))))

        (cnd? ast)
        (cons 'if (doall (map source (ast-subx ast))))

        (prim? ast)
        (cons (prim-op ast) (doall (map source (ast-subx ast))))

        (app? ast)
        (if (lam? (car (ast-subx ast)))
          (list 'let
                (doall (map (fn [p a]
                              (list (var-uid p) (source a)))
                            (lam-params (car (ast-subx ast)))
                            (cdr (ast-subx ast))))
                (source (car (ast-subx (car (ast-subx ast))))))
          (doall (map source (ast-subx ast))))

        (lam? ast)
        (list 'lambda
              (doall (map var-uid (lam-params ast)))
              (source (car (ast-subx ast))))

        (seq? ast)
        (cons 'begin (map source (ast-subx ast)))

        :else
        (error "unknown ast" ast)))

File: python/codegen.py
import re
from clojure import *

def interval (n, m):
    return range (n, m + 1)
 
lambdaTodo = list()
lambdaCount = 0
globalVars = list()
 
 
def addLambda(lam):
    global lambdaTodo, lambdaCount
    i = lambdaCount
    lambdaTodo.insert(0, cons(i, lam))
    lambdaCount = i + 1
    return i
 
def cgList (asts, vrs, stackEnv, sep, cont):
    if asts:
        x = codeGen(car(asts), stackEnv)
        return cgList(cdr(asts), cdr(vrs), cons(car(vrs), stackEnv), sep, lambda code, sv: cont([x, sep, code], sv))
    else:
        return cont("", stackEnv)
 
def cgArgs (args, stackEnv):
    return cgList(args, interval(1, len(args)), stackEnv, "", lambda code, sv: code)
 
def accessVar(v, stackEnv):
    global globalVars
    if isGlobalVar(v):
        i = posInList(v, globalVars)
        return ["GLOBAL(", i, "/*", varUid(v), "*/)"]
    else:
        i = len(stackEnv) - posInList(v, stackEnv) - 1
        return ["LOCAL(", i, "/*", varUid(v), "*/)"]
 
def cg(sv, ast):
    if isLit(ast):
        val = litVal(ast)
        return [" PUSH(INT2OBJ(", val, "));"]
 
    if isRef(ast):
        var = refVar(ast)
        return [" PUSH(", accessVar(var, sv), ");"]
 
    if isSet(ast):
        var = setVar(ast)
        return [cg(sv, car(astSubx(ast))), " ", accessVar(var, sv), " = TOS();"]
 
    if isCnd(ast):
        x = [cg(sv, i) for i in astSubx(ast)]
        return [car(x), "\n if (POP()) {\n", cadr(x), "\n } else {\n", caddr(x), "\n }"]
 
    if isPrim(ast):
        args = astSubx(ast)
        op = primOp(ast)
        if op == "%=":
            return [cgArgs(args, sv), " EQ();"]
        if op == "%<":
            return [cgArgs(args, sv), " LT();"]
        if op == "%+":
            return [cgArgs(args, sv), " ADD();"]
        if op == "%-":
            return [cgArgs(args, sv), " SUB();"]
        if op == "%*":
            return [cgArgs(args, sv), " MUL();"]
        if op == "%display":
            return [cgArgs(args, sv), " DISPLAY();"]
        if op == "%halt":
            return [cgArgs(args, sv), " HALT();"]
        if op == "%closure":
            i = addLambda(car(args))
            n = len(cdr(args))
            s = ["CLOSURE(", i, ",", n, ");"]
            return [cgArgs(cdr(args), sv), " BEGIN_", s, [[" INICLO(", i, ");"] for i in reversed(interval(1,n))], " END_", s]
        if op == "%closure-ref":
            i = litVal(cadr(args))
            return [cg(sv, car(args)), " TOS() = CLOSURE_REF(TOS(),", i, ");"]
        print("Unknown primitive")
        exit()
 
    if isApp(ast):
        fnc = car(astSubx(ast))
        args = cdr(astSubx(ast))
        n = len(args)
        if isLam(fnc):
            return cgList(args, lamParams(fnc), sv,  "\n", lambda code, newSv:[code, codeGen(car(astSubx(fnc)), newSv)])
        else:
            return cgList(
                    args,
                    interval(1, n),
                    sv,
                    "\n",
                    lambda code, newSv: [code, " BEGIN_JUMP(", n, ");", [[" PUSH(LOCAL(", j + len(sv), "));"] for j in interval(0, n - 1)], " END_JUMP(", n, ");"])
 
    print("BAD CASE")
    exit()
 
def collapse(outList, codeList):
    if codeList and isinstance(codeList, list):
        if len(codeList) > 0:
            return collapse([], first(codeList)) + collapse([], rest(codeList))
        else:
            print("This should not happen")
            exit()
    else:
        x = str(codeList)
        if x == "[]":
            return '' 
        else:
            return x
            
def code2string(codeList):
    return ''.join(collapse([], codeList))

 
def codeGen(ast, sv):
    return cg(sv, ast)
 
def compileAllLambdas():
    global lambdaTodo
    if not lambdaTodo:
        return ""
    else:
        x = car(lambdaTodo)
        ast = cdr(x)
        lambdaTodo = rest(lambdaTodo)
        
        return ["case ", car(x), ":\n", codeGen(car(astSubx(ast)), [i for i in reversed(lamParams(ast))]), "\n\n", compileAllLambdas()]

def codeGenerate(ast):
    global globalVars
    globalVars = fv(ast)
    addLambda(ast)
    code = compileAllLambdas()
    x = "#define NB_GLOBALS " + str(len(globalVars)) + "\n"
    y = "#define MAX_STACK 100 \n\n"

    code = code2string(code)

    with open("compiler/runtime.c") as f: rt = f.read()
    nrt = re.sub("//__SCHEME_CODE__", code, rt);

    return x + y + nrt
